Internal/svTrainNN.ecl

```diff
@@ -0,0 +1,494 @@
+            uint32_t * context = (uint32_t *)rtlMalloc(contxLen);
+            fT_[fT_count_].contx = context;
+            pos += sizeof(uint32_t);
+            // Copy the context wordIds into the array
+            memcpy((byte *)context, pos, contxLen);
+            rtlReleaseRow(next);
+            fT_count_++;
+          }
+          // Now we have the weights and the training pairs.  Create a neural network
+          // and train it on this data.
+          nn_ = new NeurNet(fW_, fU_, numwords, dim, lr, negsamp);
+          nn_->Train(fT_, fT_count_);
+          // Calculate the overall loss for this batch
+          loss_ = nn_->getLoss();
+          // Delete the neural network.  We're done with it.
+          delete nn_;
+          // Mark so we know we've already calculated the results.
+          calculated_ = true;
+        } // end if (!calculated)
+        // Now return the next slice.
+        // Copy the weight updates into slices to return.
+        uint32_t returnSize;
+        byte * row;
+        RtlDynamicRowBuilder rowBuilder(resultAllocator);
+        if ((outSlice_ + 1) * slicesize <= weightCount_)
+        {
+          uint32_t returnSize = sizeof(uint16_t)*2 + sizeof(float) + sizeof(float) + sizeof(uint32_t) +
+                                  sizeof(uint32_t) + sizeof(uint64_t) + sizeof(bool) + sizeof(uint32_t) +
+                                  slicesize * sizeof(double);
+          // Create a return row.  Need to use ensureCapacity for variable lenghth
+          // data.
+          row = rowBuilder.ensureCapacity(returnSize, NULL);
+          byte *pos = row;
+          *(uint16_t *)pos = nodeId_;  // nodeId (pass through)
+          pos += sizeof(uint16_t);
+          *(uint16_t *)pos = outSlice_ + 1; // Slice Id
+          pos += sizeof(uint16_t);
+          *(float *)pos = loss_; // Computed Loss
+          pos += sizeof(float);
+          *(float *)pos = minLoss_; // Min Loss (pass through)
+          pos += sizeof(float);
+          *(uint32_t *)pos = minEpoch_; // Min Epoch (pass through)
+          pos += sizeof(uint32_t);
+          // Pass through the maxNoProg
+          *(uint32_t *)pos = maxNoProg_;
+          pos += sizeof(uint32_t);
+          // Pass through the current batch position
+          *(uint64_t *)pos = batchPos_;
+          pos += sizeof(uint64_t);
+          *(bool *)pos = (bool)false; // All flag
+          pos += sizeof(bool);
+          *(uint32_t *)pos = (uint32_t)slicesize * sizeof(double); // Length of SET
+          pos += sizeof(uint32_t);
+          // Copy the weight updates into the row.
+          memcpy(pos, (byte*)fU_ + outSlice_ * slicesize * sizeof(double), slicesize * sizeof(double));
+          outSlice_++; // Increment slice number.
+          return rowBuilder.finalizeRowClear(returnSize);
+        }
+        else
+        {
+          // No more slices to output.
+          return NULL;
+        }
+    }
+    virtual void stop() override
+    {
+    }
```

Comment:
did we ever resolve whether this should call stop on its inputs.  I think it should but I know that caused issues in thor.

Reply:
@ghalliday  I think we resolved that it would be okay to leave out.  Taking it out got rid of the problem.

Reply:
temporarily removes the problem, but is not the long term fix.

Reply:
I think Jake has fixed the problem in thor which was caused when it was called twice.

Reply:
I'm going to defer fixing this to avoid a dependency on 7.2